dipy/reconst/ivim.py

```diff
@@ -0,0 +1,391 @@
+        x0 = np.array([data[0], f_guess, 10 * D_guess, D_guess])
+
+        if self.bounds is None:
+            bounds_check = [np.inf, 1., 0.1, 0.1]
+        else:
+            bounds_check = self.bounds
+
+        x0 = np.where(x0 > bounds_check[0], x0, bounds_check[0])
+        x0 = np.where(x0 < bounds_check[1], x0, bounds_check[1])
+
+        return x0
+
+    def _estimate_S0_D(self, data):
+        """
+        Obtain initial guess for S0 and D for two stage fitting.
+
+        Using TensorModel from reconst.dti, we fit an exponential
+        for those signals which are greater than a particular bvalue
+        (split_b). The apparent diffusion coefficient gives us an
+        initial estimate for D and S0.
+
+        Parameters
+        ----------
+        data : array
+            The measured signal from one voxel.
+
+        Returns
+        -------
+        S0_hat : float
+            Initial S0 guess for this voxel.
+
+        D_guess : float
+            Initial D_guess for this voxel.
+        """
+        gtab = self.gtab
+        split_b = self.split_b
+        bvals_ge_split = gtab.bvals[gtab.bvals > split_b]
+        bvecs_ge_split = gtab.bvecs[gtab.bvals > split_b]
+        gtab_ge_split = gradient_table(bvals_ge_split, bvecs_ge_split.T)
+        tensor_model = TensorModel(gtab_ge_split)
+        tenfit = tensor_model.fit(data[..., gtab.bvals > split_b])
+        D_guess = mean_diffusivity(tenfit.evals)
+        dti_params = tenfit.model_params
+        evecs = dti_params[..., 3:12].reshape((dti_params.shape[:-1] + (3, 3)))
+        evals = dti_params[..., :3]
+        qform = vec_val_vect(evecs, evals)
+
+        sphere = Sphere(xyz=gtab.bvecs[~gtab.b0s_mask])
+        ADC = apparent_diffusion_coef(qform, sphere)
+
+        S0_hat = np.mean(data[..., ~gtab.b0s_mask] /
+                         np.exp(-gtab.bvals[~gtab.b0s_mask] * ADC),
+                         -1)
+        return np.array([S0_hat, D_guess])
+
+    def _leastsq(self, data, x0):
+        """
+        Use leastsq for finding ivim_params
+
+        Parameters
+        ----------
+        data : array, (len(bvals))
+            An array containing the signal from a voxel.
+            If the data was a 3D image of 10x10x10 grid with 21 bvalues,
+            the multi_voxel decorator will run the single voxel fitting
+            on all the 1000 voxels to get the parameters in
+            IvimFit.model_paramters. The shape of the parameter array
+            will be (data[:-1], 4).
+
+        x0 : array
+            Initial guesses for the parameters S0, f, D_star and D
+            calculated using the function `_estimate_S0_D`
+        """
+        gtol = self.options["gtol"]
+        ftol = self.options["ftol"]
+        xtol = self.tol
+        epsfcn = self.options["eps"]
+        maxfev = self.options["maxiter"]
+        bounds = self.bounds
+        bvals = self.gtab.bvals
+        if not SCIPY_LESS_0_17:
```

Comment:
I think it would be more clear if you swap the order and just say `if SCIPY_LESS_0_17:` So you avoid the `if not`